File: app/src/main/java/com/leon/patmore/life/efficiency/MainActivity.java
package com.leon.patmore.life.efficiency;

import android.os.Bundle;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;

import com.leon.patmore.life.efficiency.client.LifeEfficiencyClient;
import com.leon.patmore.life.efficiency.client.LifeEfficiencyClientConfiguration;
import com.leon.patmore.life.efficiency.views.ActiveView;
import com.leon.patmore.life.efficiency.views.AddPurchaseView;
import com.leon.patmore.life.efficiency.views.GoalsView;
import com.leon.patmore.life.efficiency.views.HistoryView;
import com.leon.patmore.life.efficiency.views.MultiViewManager;
import com.leon.patmore.life.efficiency.views.RepeatingItemView;
import com.leon.patmore.life.efficiency.views.ShoppingListView;
import com.leon.patmore.life.efficiency.views.TodoHistoryView;
import com.leon.patmore.life.efficiency.views.TodoListView;
import com.leon.patmore.life.efficiency.views.WeeklyTodoView;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;

public class MainActivity extends AppCompatActivity {

    private static final String ADD_PURCHASE_VIEW_NAME = "add_purchase";
    private static final String SHOPPING_LIST_VIEW_NAME = "shopping_list";
    private static final String TODAYS_ITEMS_VIEW_NAME = "todays_items";
    private static final String REPEATING_ITEMS_VIEW_NAME = "repeating_items";
    private static final String HISTORY_VIEW_NAME = "history";

    private TodayItemsViewManager todayItemsViewManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setupViews();
        setupViewManager();
    }

    private void setupViewManager() {
        LifeEfficiencyClient lifeEfficiencyClient = LifeEfficiencyClientConfiguration.getLifeEfficiencyClientInstance();
        AutoCompleteService autoCompleteService = new AutoCompleteService(lifeEfficiencyClient);
        LinearLayout todaysItemLayout = findViewById(R.id.TodaysItemsLayout);
        Map<String, ActiveView> viewManagerMap = new HashMap<>();
        viewManagerMap.put(SHOPPING_LIST_VIEW_NAME, new ShoppingListView(findViewById(R.id.ShoppingListLayout),
                findViewById(R.id.ShoppingListButton),
                findViewById(R.id.AddToShoppingListItemName),
                findViewById(R.id.AddToShoppingListItemQuantity),
                findViewById(R.id.AddToShoppingListButton),
                findViewById(R.id.ShoppingList),
                getApplicationContext(),
                lifeEfficiencyClient,
                autoCompleteService));
        viewManagerMap.put("todoListView", new TodoListView(findViewById(R.id.TodoListLayout),
                findViewById(R.id.TodoListButton),
                findViewById(R.id.AddTodoButton),
                findViewById(R.id.AddTodoText),
                findViewById(R.id.TodoList),
                getApplicationContext(),
                lifeEfficiencyClient));
        viewManagerMap.put("todoHistoryView", new TodoHistoryView(findViewById(R.id.TodoHistoryLayout),
                findViewById(R.id.TodoHistoryButton),
                findViewById(R.id.TodoHistoryList),
                lifeEfficiencyClient,
                getApplicationContext()));
        viewManagerMap.put(TODAYS_ITEMS_VIEW_NAME, new ActiveView(todaysItemLayout, findViewById(R.id.TodaysItemsButton)) {
            @Override
            public void onActive() {
                try {
                    todayItemsViewManager.refreshList();
                } catch (ExecutionException | InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });
        viewManagerMap.put(REPEATING_ITEMS_VIEW_NAME, new RepeatingItemView(findViewById(R.id.RepeatingItemsLayout),
                findViewById(R.id.RepeatingItemsButton),
                findViewById(R.id.AddRepatingItemButton),
                findViewById(R.id.AddRepeatingItemItem),
                findViewById(R.id.RepeatingItemsList),
                lifeEfficiencyClient,
                autoCompleteService,
                getApplicationContext()));
        viewManagerMap.put(HISTORY_VIEW_NAME, new HistoryView(findViewById(R.id.HistoryLayout),
                findViewById(R.id.historyViewButton),
                lifeEfficiencyClient,
                getApplicationContext()));
        viewManagerMap.put(ADD_PURCHASE_VIEW_NAME, new AddPurchaseView(findViewById(R.id.AddPurchaseLayout),
                findViewById(R.id.AddPurchaseButton),
                lifeEfficiencyClient,
                autoCompleteService));
        viewManagerMap.put("weeklyTodo", new WeeklyTodoView(findViewById(R.id.WeeklyTodoLayout),
                findViewById(R.id.WeeklyTodoButton),
                findViewById(R.id.WeeklyTodoList),
                getApplicationContext(),
                lifeEfficiencyClient));
        viewManagerMap.put("goals", new GoalsView(findViewById(R.id.GoalsLayout),
                findViewById(R.id.GoalsButton),
                lifeEfficiencyClient,
                getApplicationContext(),
                findViewById(R.id.GoalsList)));
        new MultiViewManager(findViewById(R.id.ButtonMenuLayout),
                findViewById(R.id.BackButton),
                viewManagerMap);
    }

    private void setupViews() {
        try {
            todayItemsViewManager = new TodayItemsViewManager(findViewById(R.id.ShoppingItemList),
                    getApplicationContext(),
                    findViewById(R.id.ConfirmTodayButton));
        } catch (ExecutionException | InterruptedException e) {
            System.err.println("Could not create today's item view manager!");
            e.printStackTrace();
        }
    }

}

File: app/src/main/kotlin/com/leon/patmore/life/efficiency/MainActivity.kt
package com.leon.patmore.life.efficiency

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.leon.patmore.life.efficiency.client.LifeEfficiencyClient
import com.leon.patmore.life.efficiency.client.LifeEfficiencyClientConfiguration
import com.leon.patmore.life.efficiency.views.GoalsViewManager
import com.leon.patmore.life.efficiency.views.shopping.ShoppingAddPurchaseViewManager
import com.leon.patmore.life.efficiency.views.shopping.ShoppingHistoryViewManager
import com.leon.patmore.life.efficiency.views.shopping.ShoppingListViewManager
import com.leon.patmore.life.efficiency.views.shopping.ShoppingRepeatingItemsViewManager
import com.leon.patmore.life.efficiency.views.shopping.todays.items.ShoppingTodaysItemsViewManager
import com.leon.patmore.life.efficiency.views.todo.TodoHistoryViewManager
import com.leon.patmore.life.efficiency.views.todo.TodoListViewManager
import com.leon.patmore.life.efficiency.views.todo.TodoWeeklyViewManager

class MainActivity : AppCompatActivity() {
    private val lifeEfficiencyClient: LifeEfficiencyClient = LifeEfficiencyClientConfiguration.getLifeEfficiencyClientInstance()
    private val autoCompleteService: AutoCompleteService = AutoCompleteService(lifeEfficiencyClient)

    private val shoppingListViewData =
        ViewData(
            buttonId = R.id.ShoppingListButton,
            layoutId = R.layout.view_shopping_list,
            viewManager = ShoppingListViewManager(lifeEfficiencyClient, autoCompleteService),
        )
    private val shoppingTodaysItemsViewData =
        ViewData(
            buttonId = R.id.TodaysItemsButton,
            layoutId = R.layout.view_shopping_todays_items,
            viewManager = ShoppingTodaysItemsViewManager(lifeEfficiencyClient),
        )
    private val shoppingAddPurchaseViewData =
        ViewData(
            buttonId = R.id.AddPurchaseButton,
            layoutId = R.layout.view_shopping_add_purchase,
            viewManager = ShoppingAddPurchaseViewManager(lifeEfficiencyClient, autoCompleteService),
        )
    private val shoppingHistoryViewData =
        ViewData(
            buttonId = R.id.ShoppingHistoryButton,
            layoutId = R.layout.view_shopping_history,
            viewManager = ShoppingHistoryViewManager(lifeEfficiencyClient),
        )
    private val shoppingRepeatingItemsViewData =
        ViewData(
            buttonId = R.id.ShoppingRepeatingItemsButton,
            layoutId = R.layout.view_shopping_repeating_items,
            viewManager = ShoppingRepeatingItemsViewManager(lifeEfficiencyClient, autoCompleteService),
        )

    private val shoppingMenuSubViews =
        mutableListOf(
            shoppingListViewData,
            shoppingTodaysItemsViewData,
            shoppingAddPurchaseViewData,
            shoppingHistoryViewData,
            shoppingRepeatingItemsViewData,
        )

    private val todoListViewData =
        ViewData(
            buttonId = R.id.TodoListButton,
            layoutId = R.layout.view_todo_list,
            viewManager = TodoListViewManager(lifeEfficiencyClient),
        )
    private val todoWeeklyViewData =
        ViewData(
            buttonId = R.id.TodoWeeklyButton,
            layoutId = R.layout.view_todo_weekly,
            viewManager = TodoWeeklyViewManager(lifeEfficiencyClient),
        )
    private val todoHistoryViewData =
        ViewData(
            buttonId = R.id.TodoHistoryButton,
            layoutId = R.layout.view_todo_history,
            viewManager = TodoHistoryViewManager(lifeEfficiencyClient),
        )
    private val todoMenuSubViews = mutableListOf(todoListViewData, todoWeeklyViewData, todoHistoryViewData)

    private val goalSubView = ViewData(R.id.GoalsButton, R.layout.view_goals, viewManager = GoalsViewManager(lifeEfficiencyClient))

    private val viewButtons =
        mutableListOf(
            ViewData(R.id.ShoppingButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
            ViewData(R.id.TodoButton, R.layout.view_todo_menu, subViews = todoMenuSubViews),
            goalSubView,
        )

    init {
        shoppingMenuSubViews.add(ViewData(R.id.ShoppingMenuBackButton, R.layout.view_main_menu, subViews = viewButtons))
        shoppingTodaysItemsViewData.subViews.add(
            ViewData(R.id.ShoppingTodaysItemsBackButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
        )
        shoppingListViewData.subViews.add(
            ViewData(R.id.ShoppingListBackButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
        )
        shoppingAddPurchaseViewData.subViews.add(
            ViewData(R.id.ShoppingAddPurchaseBackButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
        )
        shoppingHistoryViewData.subViews.add(
            ViewData(R.id.ShoppingHistoryBackButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
        )
        shoppingRepeatingItemsViewData.subViews.add(
            ViewData(R.id.ShoppingRepeatingItemsBackButton, R.layout.view_shopping_menu, subViews = shoppingMenuSubViews),
        )

        todoMenuSubViews.add(ViewData(R.id.TodoMenuBackButton, R.layout.view_main_menu, subViews = viewButtons))
        todoListViewData.subViews.add(ViewData(R.id.TodoListBackButton, R.layout.view_todo_menu, subViews = todoMenuSubViews))
        todoWeeklyViewData.subViews.add(ViewData(R.id.TodoWeeklyBackButton, R.layout.view_todo_menu, subViews = todoMenuSubViews))
        todoHistoryViewData.subViews.add(ViewData(R.id.TodoHistoryBackButton, R.layout.view_todo_menu, subViews = todoMenuSubViews))
        goalSubView.subViews.add(ViewData(R.id.GoalsBackButton, R.layout.view_main_menu, subViews = viewButtons))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (todoWeeklyViewData.viewManager as TodoWeeklyViewManager).sharedPreferences =
            getPreferences(
                MODE_PRIVATE,
            )
        DataBindingUtil.setContentView<ViewDataBinding>(this, R.layout.view_main_menu)
        setupButtons(*viewButtons.toTypedArray())
    }

    private fun setupButtons(vararg viewDataList: ViewData) {
        viewDataList.forEach { viewData ->
            val view = findViewById<View>(viewData.buttonId)!!
            val button = if (view is Button) view else view.findViewById(R.id.Button)
            button.setOnClickListener {
                val binding = DataBindingUtil.setContentView<ViewDataBinding>(this, viewData.layoutId)
                setupButtons(*viewData.subViews.toTypedArray())
                if (viewData.viewManager != null) {
                    viewData.viewManager.view = binding.root
                    viewData.viewManager.context = applicationContext
                    viewData.viewManager.onActive()
                }
            }
        }
    }
}

data class ViewData(
    val buttonId: Int,
    val layoutId: Int,
    val viewManager: ViewManager? = null,
    val subViews: MutableList<ViewData> = mutableListOf(),
)

abstract class ViewManager {
    var view: View? = null
    var context: Context? = null

    abstract fun onActive()
}
